### PR TITLE
[Cache] Fix Predis client cluster with pipeline

### PR DESCRIPTION
**Pull request by flolivaud**
| Q             | A
| ------------- | ---
| Branch?       | 3.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | n/a

### CODE REVIEW

File: src/Symfony/Component/Cache/Traits/RedisTrait.php [hunk 1]
```diff
@@ -284,7 +285,7 @@ private function pipeline(\Closure $generator)
     {
         $ids = array();
 
-        if ($this->redis instanceof \Predis\Client) {
+        if ($this->redis instanceof \Predis\Client && !($this->redis->getConnection() instanceof ClusterInterface)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Traits/RedisTrait.php]: No need for the parentheses.

File: src/Symfony/Component/Cache/Traits/RedisTrait.php [hunk 2]
```diff
@@ -308,9 +309,12 @@ private function pipeline(\Closure $generator)
             foreach ($results as $k => list($h, $c)) {
                 $results[$k] = $connections[$h][$c];
             }
-        } elseif ($this->redis instanceof \RedisCluster) {
-            // phpredis doesn't support pipelining with RedisCluster
+        } elseif ($this->redis instanceof \RedisCluster || (
+                $this->redis instanceof \Predis\Client && $this->redis->getConnection() instanceof ClusterInterface
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Traits/RedisTrait.php]: Should be on one​ line.

**Review by nicolas-grekas — APPROVED** [review]
:+1: with minor comments

**Comment by fabpot** [comment]
@flolivaud Can you change the base to the lowest possible version?

**Comment by flolivaud** [comment]
@fabpot Done.

**Comment by keradus** [comment]
when lowering the branch on gh interface, one needs to rebase the branch as well, as if not, he ends up with all commits from original base branch in his PR

**Comment by flolivaud** [comment]
@keradus yes sorry, seems to be good ! /cc @fabpot

**Comment by keradus** [comment]
all bugfixes deserve a test. please add one

**Comment by nicolas-grekas** [comment]
Note that in this case, having a test might be complicated...

**Comment by flolivaud** [comment]
Yes it is. I fixed the test to connect in cluster mode but i don't know how to do more.

File: src/Symfony/Component/Cache/Traits/RedisTrait.php [hunk 3]
```diff
@@ -308,9 +309,10 @@ private function pipeline(\Closure $generator)
             foreach ($results as $k => list($h, $c)) {
                 $results[$k] = $connections[$h][$c];
             }
-        } elseif ($this->redis instanceof \RedisCluster) {
-            // phpredis doesn't support pipelining with RedisCluster
+        } elseif ($this->redis instanceof \RedisCluster || ($this->redis instanceof \Predis\Client && $this->redis->getConnection() instanceof ClusterInterface)) {
+            // phpredis & predis doesn't support pipelining with RedisCluster
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Traits/RedisTrait.php]: s/doesn't/don't

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/Cache/Traits/RedisTrait.php (hunk 3 above)
Comment (flolivaud) [inline src/Symfony/Component/Cache/Traits/RedisTrait.php]: Sorry. Done. :)

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @flolivaud.

**Comment by flolivaud** [comment]
@fabpot the fix is actually on the 3.3 branch, but need to be on 3.3+ branches (master included).

**Comment by xabbuh** [comment]
@flolivaud branches are merged up regularly so each fix will eventually land in all still maintained branches

**Pull request merged**